File: ios/ParleyKit/Sources/ParleyKit/CaptureRecovery.swift
import Foundation

/// When to try to take the microphone back, how long to keep trying, and when to
/// stop pretending — as a state machine rather than as retries scattered through
/// the capture.
///
/// ## The failure this exists for
///
/// The app holds the microphone in the background (a `MicWindowState` window, or
/// a dictation the user swiped away from). Something with a higher claim on the
/// input starts: the system keyboard's dictation, Siri, a call. iOS posts
/// `AVAudioSession.interruptionNotification` `.began`, stops our IO, and the
/// recovery that used to follow made two assumptions that are both wrong in
/// exactly that situation:
///
/// 1. **That `.ended` arrives.** System services do not always post it. An
///    interruption with no end left the capture parked on `interrupted`
///    forever — and `interrupted` was also what disabled the watchdog, so the
///    one backstop for "iOS announced nothing" could not see the one case it
///    was built for.
/// 2. **That trying six times is trying.** The ladder ran out in under twelve
///    seconds. Every one of those attempts was `AVAudioSession.setActive(true)`
///    from a backgrounded app that another client had just interrupted, which
///    iOS refuses (`!pri`, `!int`, `!rec`) until the app is in the foreground
///    again. Twelve seconds of being refused, and then a capture that was
///    inert for good: nothing in the process watched for the foreground, or for
///    `mediaServicesWereReset`, so the two moments where reactivation would
///    have worked went by unnoticed.
///
/// So: probe even while interrupted (the only way to notice a missing
/// `.ended`), climb a ladder long enough to outlast a dictation rather than a
/// route change, and treat *giving up as a state that is still armed* —
/// `.lost` answers `appBecameActive` and `mediaServicesReset` with one more
/// rebuild, because those are the two events that change the answer.
///
/// Giving up is not silence. `.giveUp` is the cue to tell the keyboard that the
/// microphone is gone, which is the difference between a pane that has stopped
/// listening and a pane that only looks like it is listening.
///
/// Pure logic, no AVFoundation: the sequence that produces the bug takes a
/// minute of real time on a real phone and is three lines in a test.
public struct CaptureRecovery: Sendable, Equatable {

    /// Everything that can change the answer.
    public enum Event: Sendable, Equatable {
        /// `AVAudioSession` interruption `.began` — the system has our input.
        case interrupted
        /// `AVAudioSession` interruption `.ended`. Advisory `.shouldResume` is
        /// deliberately not part of this: a recording the user never stopped
        /// always wants the microphone back.
        case interruptionEnded
        /// A rebuild attempt opened the engine.
        case rebuildSucceeded
        /// A rebuild attempt threw. `systemHoldsInput` separates "somebody else
        /// has the input, or we are backgrounded and not allowed to take it"
        /// from "this device's audio is broken", which is the difference between
        /// a loss worth offering a restart for and one worth naming.
        case rebuildFailed(systemHoldsInput: Bool, description: String)
        /// The app came to the foreground. The one state change that reliably
        /// turns a refused `setActive(true)` into an allowed one.
        case appBecameActive
        /// `mediaserverd` restarted: every audio object in the process is a
        /// corpse, and the session the other client was holding died with it.
        case mediaServicesReset
        /// The engine can no longer be trusted for a reason that is ours rather
        /// than the system's — a route change, a configuration change, the
        /// watchdog noticing it stopped. Nobody is holding the input, so there
        /// is nothing to wait for.
        case engineStopped
        /// A fresh `start()` — the user tapped. Everything this policy was
        /// remembering belongs to the capture that is being replaced.
        case restarted
    }

    /// What the capture should do about it.
    public enum Action: Sendable, Equatable {
        /// Nothing, and in particular not a rebuild: either the engine is fine
        /// or a chain is already climbing the ladder.
        case wait
        /// Rebuild the engine after this delay. Zero means now.
        case rebuild(afterMilliseconds: Int)
        /// The ladder ran out. Tell the keyboard, and stay armed.
        case giveUp(Loss)
    }

    /// Why the microphone is gone, in the only terms a user can act on.
    public enum Loss: Sendable, Equatable {
        /// Something else has the input, or iOS will not let a backgrounded app
        /// take it back. A trip through the foreground fixes it, which is what
        /// the keyboard's copy offers.
        case takenBySystem
        /// The audio stack itself refused, with a reason worth repeating.
        case broken(String)
    }

    public enum Phase: Sendable, Equatable {
        /// The microphone is ours.
        case running
        /// `.began` arrived and `.ended` has not. Probes run anyway.
        case interrupted
        /// A rebuild chain is climbing the ladder.
        case recovering
        /// The ladder ran out and the keyboard has been told. Still armed: a
        /// foreground trip or a media-services reset starts a fresh chain.
        case lost
    }

    /// How long to wait before probing an interruption that has not announced
    /// its end. Long enough not to snatch the input back from whatever is
    /// mid-sentence in it, short enough that a dictation the user cancelled
    /// after a second is not paid for in ten.
    public var probeAfterMilliseconds: Int
    /// How long to wait after `.ended`. The route is still settling; the input
    /// node reports 0 Hz for a beat if asked immediately.
    public var resumeAfterMilliseconds: Int
    public var firstBackoffMilliseconds: Int
    public var backoffCapMilliseconds: Int
    /// Attempts before the keyboard is told. Twelve at this ladder is roughly
    /// half a minute of trying — long enough to outlast a system dictation,
    /// where six attempts (under twelve seconds) could only ever outlast a
    /// route change.
    public var maxAttempts: Int

    public private(set) var phase: Phase = .running
    /// Consecutive failed rebuilds in the current chain.
    public private(set) var attempts = 0
    /// The chain started with the system taking the input. Kept because the
    /// *last* error need not be the one that explains the situation — a probe
    /// refused while Siri holds the mic can surface as a 0 Hz input node — and
    /// what the user is told has to describe what happened, not the last thing
    /// that went wrong.
    private var beganWithInterruption = false

    public init(
        probeAfterMilliseconds: Int = 2_500,
        resumeAfterMilliseconds: Int = 250,
        firstBackoffMilliseconds: Int = 250,
        backoffCapMilliseconds: Int = 4_000,
        maxAttempts: Int = 12
    ) {
        self.probeAfterMilliseconds = probeAfterMilliseconds
        self.resumeAfterMilliseconds = resumeAfterMilliseconds
        self.firstBackoffMilliseconds = firstBackoffMilliseconds
        self.backoffCapMilliseconds = backoffCapMilliseconds
        self.maxAttempts = maxAttempts
    }

    /// The microphone is not ours right now, whether or not anyone has been
    /// told yet. `AudioCapture` reports this as "not capturing", which is what
    /// stops the next dictation from borrowing a capture that cannot hear.
    public var holdsMicrophone: Bool { phase == .running }

    /// The keyboard has been told. Still armed — see `Phase.lost`.
    public var hasGivenUp: Bool { phase == .lost }

    /// `1×, 2×, 4×, …` the first backoff, capped. Shift rather than `pow` for
    /// the same reason `ReconnectPolicy` uses one: exact, and it cannot drift
    /// past the cap through floating point.
    public func backoff(forAttempt attempt: Int) -> Int {
        guard attempt >= 1 else { return firstBackoffMilliseconds }
        let steps = min(attempt - 1, 30)
        let scaled = firstBackoffMilliseconds << steps
        return min(scaled, backoffCapMilliseconds)
    }

    /// Total time the ladder spends trying before the keyboard is told,
    /// counting from the first failure. Exists for the test that keeps this
    /// honest about outlasting a system dictation.
    public var ladderMilliseconds: Int {
        guard maxAttempts > 1 else { return 0 }
        return (1..<maxAttempts).reduce(0) { $0 + backoff(forAttempt: $1) }
    }

    public mutating func apply(_ event: Event) -> Action {
        switch event {
        case .restarted:
            phase = .running
            attempts = 0
            beganWithInterruption = false
            return .wait

        case .rebuildSucceeded:
            phase = .running
            attempts = 0
            beganWithInterruption = false
            return .wait

        case .interrupted:
            beganWithInterruption = true
            // A second `.began` for an interruption already being handled says
            // nothing new, and answering it with another probe would only put a
            // second chain on the same engine.
            guard phase == .running else { return .wait }
            phase = .interrupted
            attempts = 0
            // A probe rather than a wait. This is the *only* thing that can
            // notice an interruption whose `.ended` never comes — and being
            // refused while the other client still holds the input costs one
            // throw, which is what the ladder is for.
            return .rebuild(afterMilliseconds: probeAfterMilliseconds)

        case .interruptionEnded:
            // From `.lost` too: the system announcing that it is done with the
            // microphone is the best reason there has ever been to try again,
            // whatever this policy had concluded before it.
            phase = .recovering
            attempts = 0
            return .rebuild(afterMilliseconds: resumeAfterMilliseconds)

        case .appBecameActive:
            // The refusal that burned the ladder was "you are in the background
            // and somebody else is running". Being in the foreground is the
            // other half of that answer.
            guard phase != .running else { return .wait }
            phase = .recovering
            attempts = 0
            return .rebuild(afterMilliseconds: 0)

        case .mediaServicesReset:
            phase = .recovering
            attempts = 0
            return .rebuild(afterMilliseconds: resumeAfterMilliseconds)

        case .engineStopped:
            phase = .recovering
            attempts = 0
            return .rebuild(afterMilliseconds: 0)

        case .rebuildFailed(let systemHoldsInput, let description):
            // Already given up: a straggler from the chain that gave up is not
            // a reason to tell the keyboard twice.
            guard phase != .lost else { return .wait }
            attempts += 1
            if systemHoldsInput { beganWithInterruption = true }
            guard attempts < maxAttempts else {
                phase = .lost
                return .giveUp(
                    beganWithInterruption || systemHoldsInput
                        ? .takenBySystem : .broken(description))
            }
            phase = .recovering
            return .rebuild(afterMilliseconds: backoff(forAttempt: attempts))
        }
    }
}

File: android/parleykit/src/main/kotlin/com/pathors/parley/kit/CaptureRecovery.kt
package com.pathors.parley.kit

/**
 * When to try to take the microphone back, how long to keep trying, and when to
 * stop pretending — as a state machine rather than as retries scattered through
 * the capture.
 *
 * A line-for-line port of iOS
 * `ios/ParleyKit/Sources/ParleyKit/CaptureRecovery.swift`, kept in this module
 * for the same reason that one lives in ParleyKit: it is the only part of
 * microphone recovery that can be tested without a microphone. The sequence
 * that produces the bug takes a real phone, a real interruption and a minute of
 * waiting to reproduce once, and is three lines in a test.
 *
 * ## The failure this exists for
 *
 * `MicCapture` was a deliberate one-shot: open an `AudioRecord`, read until it
 * fails, end the flow. The comment on it said, in as many words, "there is no
 * restart — build a new MicCapture". Everything downstream honoured that, so a
 * phone call, Google Assistant, or another recorder starting up ended the
 * meeting. The first wave of durability work made that survivable — the file is
 * finished and uploaded instead of deleted — but survivable is not the same as
 * recovered: the recording still *stopped*, forty seconds into an interruption
 * that lasted four.
 *
 * Android gives us three separate ways to learn about it and none of them is
 * reliable on its own:
 *
 * 1. **`AudioRecord.read` returns a negative code.** `ERROR_DEAD_OBJECT` or
 *    `ERROR_INVALID_OPERATION`. Honest, immediate — and only some devices do
 *    it.
 * 2. **`AudioRecordingCallback.isClientSilenced`.** From Android 10 an app that
 *    loses the microphone to a higher-priority client is fed *silence* rather
 *    than an error, and this callback is the only announcement. It is also the
 *    only one of the three that can fire while `read()` keeps happily returning
 *    success.
 * 3. **Nothing at all.** `read()` stops returning, or returns zeroes with no
 *    callback. This is what the watchdog is for.
 *
 * So the same two wrong assumptions iOS had to unlearn apply here:
 *
 * - **That the end of an interruption is announced.** It frequently is not:
 *   `isClientSilenced` can go true and never come back, because the other
 *   client released the microphone without the framework re-running our
 *   callback. Waiting for an "ended" that never arrives parks the capture
 *   forever.
 * - **That trying a handful of times is trying.** A short ladder runs out
 *   inside a route change. An in-call recording, a system dictation or another
 *   app's recorder holds the microphone for as long as the *user* is busy with
 *   it, which is tens of seconds, not five.
 *
 * Hence: probe even while interrupted (the only way to notice a missing end),
 * climb a ladder long enough to outlast an interruption rather than a route
 * change, and treat *giving up as a state that is still armed* — [Phase.LOST]
 * answers [Event.AppBecameActive] and [Event.AudioServerDied] with one more
 * rebuild, because those are the two events that change the answer.
 *
 * Giving up is not silence either. [Action.GiveUp] is the cue to tell the user
 * the microphone is gone, which is the difference between a meeting that has
 * stopped listening and one that only looks like it is still listening. It is
 * emphatically *not* a cue to throw the recording away: see
 * `meeting/CaptureEnding.kt` for why exactly one ending in this app deletes
 * audio and why it is the one a person asked for.
 *
 * Synchronized throughout, like `ReconnectPolicy`: the `AudioRecord` read
 * thread, the device-routing callback, the silenced-client callback, the
 * watchdog thread and the activity-lifecycle callback all reach this from
 * whichever thread the platform handed them.
 *
 * @property probeAfterMillis how long to wait before probing an interruption
 *   that has not announced its end. Long enough not to snatch the microphone
 *   back from whatever is mid-sentence in it, short enough that an interruption
 *   the user cancelled after a second is not paid for in ten.
 * @property resumeAfterMillis how long to wait after the interruption ends. The
 *   route is still settling; `AudioRecord.Builder` can still fail for a beat
 *   after the other client has let go.
 * @property maxAttempts attempts before the user is told. Twelve at this ladder
 *   is roughly half a minute of trying — long enough to outlast a call being
 *   answered and declined, where six attempts (under twelve seconds) could only
 *   ever outlast a route change.
 */
class CaptureRecovery(
    val probeAfterMillis: Long = DEFAULT_PROBE_AFTER_MILLIS,
    val resumeAfterMillis: Long = DEFAULT_RESUME_AFTER_MILLIS,
    val firstBackoffMillis: Long = DEFAULT_FIRST_BACKOFF_MILLIS,
    val backoffCapMillis: Long = DEFAULT_BACKOFF_CAP_MILLIS,
    val maxAttempts: Int = DEFAULT_MAX_ATTEMPTS,
) {
    init {
        require(probeAfterMillis >= 0) { "probeAfterMillis must not be negative" }
        require(resumeAfterMillis >= 0) { "resumeAfterMillis must not be negative" }
        require(firstBackoffMillis > 0) {
            "firstBackoffMillis must be positive (got $firstBackoffMillis)"
        }
        require(backoffCapMillis >= firstBackoffMillis) {
            "backoffCapMillis ($backoffCapMillis) must not be shorter than " +
                "firstBackoffMillis ($firstBackoffMillis)"
        }
        require(maxAttempts >= 1) { "maxAttempts must be at least 1 (got $maxAttempts)" }
    }

    /** Everything that can change the answer. */
    sealed interface Event {
        /**
         * Something with a higher claim has our input: `isClientSilenced` went
         * true, or `AudioRecord.read` returned an error. The Android equivalent
         * of iOS's `AVAudioSession` interruption `.began`.
         */
        data object Interrupted : Event

        /**
         * The interruption is over — `isClientSilenced` went false again.
         *
         * There is deliberately no "should resume" advisory in this event, for
         * the same reason iOS does not consult `AVAudioSession`'s: a recording
         * the user never stopped always wants the microphone back.
         */
        data object InterruptionEnded : Event

        /** A rebuild attempt opened and started an `AudioRecord`. */
        data class RebuildSucceeded(val sampleRate: Int = 0) : Event

        /**
         * A rebuild attempt failed.
         *
         * [systemHoldsInput] separates "somebody else has the microphone" —
         * `AudioRecord.Builder` throwing, `state != STATE_INITIALIZED`, or
         * `recordingState != RECORDSTATE_RECORDING`, all of which are what a
         * busy microphone looks like — from "this device's audio is broken",
         * which is the difference between a loss worth offering a retry for and
         * one worth naming.
         */
        data class RebuildFailed(
            val systemHoldsInput: Boolean,
            val description: String? = null,
        ) : Event

        /**
         * The app came to the foreground. Not an audio event at all, and that
         * is exactly why it matters: the platform is markedly more willing to
         * hand a foreground app the microphone, and this is the one state
         * change that can turn a refusal into an acceptance. Delivered from
         * `Application.ActivityLifecycleCallbacks.onActivityResumed`.
         */
        data object AppBecameActive : Event

        /**
         * `AudioRecord.read` returned `ERROR_DEAD_OBJECT`: the audio server
         * restarted and every `AudioRecord` in the process is a corpse — as is
         * the one the other client was holding. The Android analogue of iOS's
         * `mediaServicesWereResetNotification`, and worth trying again after
         * for exactly the same reason: whatever was holding the microphone died
         * with it.
         */
        data object AudioServerDied : Event

        /**
         * Capture can no longer be trusted for a reason that is ours rather
         * than the system's — the input route moved to a device we are not
         * recording from, or the watchdog noticed chunks stopped arriving.
         * Nobody is holding the microphone, so there is nothing to wait for.
         */
        data object CaptureStopped : Event

        /**
         * A fresh start — the user tapped record. Everything this policy was
         * remembering belongs to the capture that is being replaced.
         */
        data object Restarted : Event
    }

    /** What the capture should do about it. */
    sealed interface Action {
        /**
         * Nothing, and in particular not a rebuild: either the `AudioRecord` is
         * fine or a chain is already climbing the ladder.
         */
        data object Wait : Action

        /** Rebuild the `AudioRecord` after this delay. Zero means now. */
        data class Rebuild(val afterMillis: Long) : Action

        /** The ladder ran out. Tell the user, and stay armed. */
        data class GiveUp(val loss: Loss) : Action
    }

    /** Why the microphone is gone, in the only terms a user can act on. */
    sealed interface Loss {
        /**
         * Something else has the microphone. Putting the phone down and coming
         * back to Parley, or ending the call, fixes it — which is what the
         * banner copy offers.
         */
        data object TakenBySystem : Loss

        /** The audio stack itself refused, with a reason worth repeating. */
        data class Broken(val description: String?) : Loss
    }

    enum class Phase {
        /** The microphone is ours. */
        RUNNING,

        /** An interruption began and has not ended. Probes run anyway. */
        INTERRUPTED,

        /** A rebuild chain is climbing the ladder. */
        RECOVERING,

        /**
         * The ladder ran out and the user has been told. Still armed: the app
         * coming forward, or the audio server restarting, starts a fresh chain.
         */
        LOST,
    }

    @get:Synchronized
    var phase: Phase = Phase.RUNNING
        private set

    /** Consecutive failed rebuilds in the current chain. */
    @get:Synchronized
    var attempts: Int = 0
        private set

    /**
     * The chain started with the system taking the microphone. Kept because the
     * *last* error need not be the one that explains the situation — a probe
     * attempted while another app holds the input can surface as an
     * unsupported-configuration failure — and what the user is told has to
     * describe what happened, not the last thing that went wrong.
     */
    private var beganWithInterruption = false

    /**
     * The microphone is not ours right now, whether or not anyone has been told
     * yet. This is what stops the rest of the app from believing a capture that
     * cannot hear.
     */
    @get:Synchronized
    val holdsMicrophone: Boolean get() = phase == Phase.RUNNING

    /** The user has been told. Still armed — see [Phase.LOST]. */
    @get:Synchronized
    val hasGivenUp: Boolean get() = phase == Phase.LOST

    /** A chain is mid-climb, which is what suppresses the watchdog and route changes. */
    @get:Synchronized
    val isRecovering: Boolean get() = phase == Phase.RECOVERING || phase == Phase.INTERRUPTED

    /**
     * `1×, 2×, 4×, …` the first backoff, capped. A shift rather than `pow`, for
     * the same reason `ReconnectPolicy` uses one: it is exact, and it cannot
     * drift past the cap through floating point.
     */
    fun backoffFor(attempt: Int): Long {
        if (attempt < 1) return firstBackoffMillis
        val steps = minOf(attempt - 1, MAX_BACKOFF_STEPS)
        return minOf(firstBackoffMillis shl steps, backoffCapMillis)
    }

    /**
     * Total time the ladder spends trying before the user is told, counting
     * from the first failure. Exists for the test that keeps this honest about
     * outlasting an interruption a person is busy with.
     */
    val ladderMillis: Long
        get() = if (maxAttempts <= 1) 0L else (1 until maxAttempts).sumOf { backoffFor(it) }

    /**
     * The event table, deliberately flat and in the same order as the Swift
     * `switch` it was ported from, so the two platforms can still be diffed by
     * eye. The three events whose answer depends on the phase we are already in
     * delegate to a named function below rather than nesting a decision inside
     * an arm.
     *
     * Those helpers write [phase] and [attempts] without taking the lock
     * themselves, and are private so that this stays true: they are only ever
     * reached through this method, which holds it.
     */
    @Synchronized
    fun apply(event: Event): Action = when (event) {
        Event.Restarted, is Event.RebuildSucceeded -> {
            phase = Phase.RUNNING
            attempts = 0
            beganWithInterruption = false
            Action.Wait
        }

        Event.Interrupted -> onInterrupted()

        Event.InterruptionEnded -> {
            // Reached from LOST too: the system announcing that it is done with
            // the microphone is the best reason there has ever been to try
            // again, whatever this policy had concluded before it.
            phase = Phase.RECOVERING
            attempts = 0
            Action.Rebuild(resumeAfterMillis)
        }

        Event.AppBecameActive -> onAppBecameActive()

        // The same answer as [Event.InterruptionEnded], and today the same
        // delay — but deliberately *not* folded into one arm with it, because
        // it is not the same fact. That one is the other client announcing it
        // has let go; this one is every `AudioRecord` in the process dying with
        // the audio server that was holding it, ours included. Two reasons to
        // keep them apart. This file's stated value is being a line-for-line
        // port of `CaptureRecovery.swift`, where they are separate cases and a
        // reader can check the two platforms agree by looking. And
        // [resumeAfterMillis] is documented as the settle after an
        // *interruption* — of the numbers here it is the likeliest to want a
        // different value once a restarted audio server has been measured,
        // which a shared arm would silently apply to both.
        Event.AudioServerDied -> { // NOSONAR — S1871: same answer, different reasons; see above
            phase = Phase.RECOVERING
            attempts = 0
            Action.Rebuild(resumeAfterMillis)
        }

        Event.CaptureStopped -> {
            phase = Phase.RECOVERING
            attempts = 0
            Action.Rebuild(0)
        }

        is Event.RebuildFailed -> onRebuildFailed(event)
    }

    /**
     * [Event.Interrupted]: probe, unless a chain is already climbing.
     *
     * [beganWithInterruption] is recorded either way. The interruption happened
     * whether or not this is the call that starts the chain, and it is what the
     * eventual [Action.GiveUp] describes the loss with.
     */
    private fun onInterrupted(): Action {
        beganWithInterruption = true
        // A second beginning for an interruption already being handled says
        // nothing new, and answering it with another probe would only put a
        // second chain on the same `AudioRecord`.
        if (phase != Phase.RUNNING) return Action.Wait
        phase = Phase.INTERRUPTED
        attempts = 0
        // A probe rather than a wait. This is the *only* thing that can notice
        // an interruption whose end is never announced — and being refused
        // while the other client still holds the microphone costs one failed
        // open, which is what the ladder is for.
        return Action.Rebuild(probeAfterMillis)
    }

    /**
     * [Event.AppBecameActive]: one immediate attempt from every phase except
     * [Phase.RUNNING], where the microphone is already ours and there is
     * nothing to take back.
     */
    private fun onAppBecameActive(): Action {
        if (phase == Phase.RUNNING) return Action.Wait
        // The refusal that burned the ladder was "a backgrounded app, competing
        // with something in the foreground". Being in the foreground is the
        // other half of that answer.
        phase = Phase.RECOVERING
        attempts = 0
        return Action.Rebuild(0)
    }

    /** [Event.RebuildFailed]: the next rung of the ladder, or the end of it. */
    private fun onRebuildFailed(event: Event.RebuildFailed): Action {
        // Already given up: a straggler from the chain that gave up is not a
        // reason to tell the user twice.
        if (phase == Phase.LOST) return Action.Wait
        attempts += 1
        if (event.systemHoldsInput) beganWithInterruption = true
        if (attempts < maxAttempts) {
            phase = Phase.RECOVERING
            return Action.Rebuild(backoffFor(attempts))
        }
        phase = Phase.LOST
        // What the user is told has to describe what happened rather than the
        // last thing that went wrong — see [beganWithInterruption].
        return Action.GiveUp(
            if (beganWithInterruption || event.systemHoldsInput) {
                Loss.TakenBySystem
            } else {
                Loss.Broken(event.description)
            },
        )
    }

    companion object {
        const val DEFAULT_PROBE_AFTER_MILLIS = 2_500L
        const val DEFAULT_RESUME_AFTER_MILLIS = 250L
        const val DEFAULT_FIRST_BACKOFF_MILLIS = 250L
        const val DEFAULT_BACKOFF_CAP_MILLIS = 4_000L
        const val DEFAULT_MAX_ATTEMPTS = 12

        /** Shift ceiling; well past where [DEFAULT_BACKOFF_CAP_MILLIS] clamps anyway. */
        private const val MAX_BACKOFF_STEPS = 30
    }
}
